File: src/com/company/Case.java
package com.company;

/**
 * Created by ArO on 27/08/2016.
 */
public class Case implements Comparable<Case>{

    protected int journeyCode;
    protected String holidayType;
    protected int holidayTypeIndex;
    protected int price;
    protected int numPerson;
    protected String region;
    protected String transportation;
    protected int transportationIndex;
    protected int duration;
    protected String season;
    protected int seasonIndex;
    protected String accommodation;
    protected int accommodationIndex;
    protected String hotel;
    protected double similarity;

    public void setupIndices(){
        this.holidayTypeIndex = getHolidayIndex(this.holidayType);
        this.transportationIndex = getTransportIndex(this.transportation);
        this.seasonIndex = getSeasonIndex(this.season);
        this.accommodationIndex = getAccomodationIndex(this.accommodation);
    }


    private int getTransportIndex(String type){
        if(type.equals("Plane")){
            return 0;
        }
        else if(type.equals("Car")){
            return 1;
        }
        else if(type.equals("Train")){
            return 2;
        }
        else{
            return 3;
        }
    }
    private int getSeasonIndex(String season){
        if(season.equals("January")){
            return 1;
        }else if(season.equals("February")){
            return 2;
        }
        else if(season.equals("March")){
            return 3;
        }
        else if(season.equals("April")){
            return 4;
        }
        else if(season.equals("May")){
            return 5;
        }
        else if(season.equals("June")){
            return 6;
        }
        else if(season.equals("July")){
            return 7;
        }
        else if(season.equals("August")){
            return 8;
        }
        else if(season.equals("September")){
            return 9;
        }
        else if(season.equals("October")){
            return 10;
        }
        else if(season.equals("November")){
            return 11;
        }
        else{
            return 12;
        }
    }
    private int getAccomodationIndex(String accomodation){
        if(accomodation.equals("HolidayFlat")){
            return 0;
        }
        else if(accomodation.equals("OneStar")){
            return 1;
        }
        else if(accomodation.equals("TwoStars")){
            return 2;
        }
        else if(accomodation.equals("ThreeStars")){
            return 3;
        }
        else if(accomodation.equals(("FourStars"))){
            return 4;
        }
        else{
            return 5;
        }
    }
    private int getHolidayIndex(String holiday){
        //First group
        if(holiday.equals("Bathing")){
            return 0;
        }else if(holiday.equals("City")){
            return 3;
        }else if(holiday.equals("Wandering")){
            return 6;
        }//Second group
        else if(holiday.equals("Active")){
            return 1;
        }else if(holiday.equals("Recreation")){
            return 4;
        }else if(holiday.equals("Skiing")){
            return 7;
        }//Third group
        else if(holiday.equals("Education")){
            return 2;
        }
        return 5;
    }

    public void printCase(){
        System.out.println("Case: " + journeyCode);
        System.out.println(holidayType);
        System.out.println(price);
        System.out.println(numPerson);
        System.out.println(region);
        System.out.println(transportation);
        System.out.println(duration);
        System.out.println(season);
        System.out.println(accommodation);
        System.out.println(hotel);
        System.out.println("similarity: " + similarity + "\n");

    }

    @Override
    public int compareTo(Case other) {
        if(this.similarity == other.similarity){
            return 0;
        }
        return this.similarity < other.similarity ? 1 : -1;
    }

    @Override
    public String toString(){
        String output = "";
        if(journeyCode != 0){
            output += "Journey Code: " + journeyCode + "\n";
        }
        output += "Holiday Type: " + holidayType + "\n";
        output += "Price: " + price + "\n";
        output += "Number of People: " + numPerson + "\n";
        output += "Region: " + region + "\n";
        output += "Transportation: " + transportation + "\n";
        output += "Duration: " + duration + "\n";
        output += "Season: " + season + "\n";
        output += "Accomodation: " + accommodation + "\n";
        if(hotel != null){
            output +="Hotel: " + hotel + "\n";
        }
        if(similarity != 0){
            output += "Similarity: " + similarity + "\n";
        }

        return output;
    }
}

File: csharp/CasedBasedReasoning/CasedBasedReasoning/Case.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CasedBasedReasoning
{
    public class Case : IComparable
    {
        public int JourneyCode;
        public string HolidayType;
        public int HolidayTypeIndex;
        public int Price;
        public int NumPerson;
        public string Region;
        public string Transportation;
        public int TransportationIndex;
        public int Duration;
        public string Season;
        public int SeasonIndex;
        public string Accommodation;
        public int AccommodationIndex;
        public string Hotel;
        public double Similarity;

        public void SetupIndices()
        {
            HolidayTypeIndex = GetHolidayIndex(HolidayType);
            TransportationIndex = GetTransportationIndex(Transportation);
            SeasonIndex = GetSeasonIndex(Season);
            AccommodationIndex = GetAccommodationIndex(Accommodation);
        }

        private int GetAccommodationIndex(string accommodation)
        {
            if (accommodation.Equals("HolidayFlat"))
            {
                return 0;
            }
            else if (accommodation.Equals("OneStar"))
            {
                return 1;
            }
            else if (accommodation.Equals("TwoStars"))
            {
                return 2;
            }
            else if (accommodation.Equals("ThreeStars"))
            {
                return 3;
            }
            else if (accommodation.Equals(("FourStars")))
            {
                return 4;
            }
            else
            {
                return 5;
            }
        }

        private int GetSeasonIndex(string season)
        {
            if (season.Equals("January"))
            {
                return 1;
            }
            else if (season.Equals("February"))
            {
                return 2;
            }
            else if (season.Equals("March"))
            {
                return 3;
            }
            else if (season.Equals("April"))
            {
                return 4;
            }
            else if (season.Equals("May"))
            {
                return 5;
            }
            else if (season.Equals("June"))
            {
                return 6;
            }
            else if (season.Equals("July"))
            {
                return 7;
            }
            else if (season.Equals("August"))
            {
                return 8;
            }
            else if (season.Equals("September"))
            {
                return 9;
            }
            else if (season.Equals("October"))
            {
                return 10;
            }
            else if (season.Equals("November"))
            {
                return 11;
            }
            else
            {
                return 12;
            }
        }

        private int GetTransportationIndex(string type)
        {
            if (type.Equals("Plane"))
            {
                return 0;
            }
            else if (type.Equals("Car"))
            {
                return 1;
            }
            else if (type.Equals("Train"))
            {
                return 2;
            }
            else
            {
                return 3;
            }
        }

        private int GetHolidayIndex(string holiday)
        {
            if (holiday.Equals("Bathing"))
            {
                return 0;
            }
            else if (holiday.Equals("City"))
            {
                return 3;
            }
            else if (holiday.Equals("Wandering"))
            {
                return 6;
            }//Second group
            else if (holiday.Equals("Active"))
            {
                return 1;
            }
            else if (holiday.Equals("Recreation"))
            {
                return 4;
            }
            else if (holiday.Equals("Skiing"))
            {
                return 7;
            }//Third group
            else if (holiday.Equals("Education"))
            {
                return 2;
            }
            return 5;
        }

        public void printCase()
        {
            Console.WriteLine("Case: " + JourneyCode);
            Console.WriteLine(HolidayType);
            Console.WriteLine(Price);
            Console.WriteLine(NumPerson);
            Console.WriteLine(Region);
            Console.WriteLine(Transportation);
            Console.WriteLine(Duration);
            Console.WriteLine(Season);
            Console.WriteLine(Accommodation);
            Console.WriteLine(Hotel);
            Console.WriteLine("similarity: " + Similarity + "\n");

        }

        public int CompareTo(Case other)
        {
            if (Similarity == other.Similarity)
            {
                return 0;
            }
            return Similarity < other.Similarity ? 1 : -1;
        }

        public override string ToString()
        {
            string output = "";
            if (JourneyCode != 0)
            {
                output += "Journey Code: " + JourneyCode + "\n";
            }
            output += "Holiday Type: " + HolidayType + "\n";
            output += "Price: " + Price + "\n";
            output += "Number of People: " + NumPerson + "\n";
            output += "Region: " + Region + "\n";
            output += "Transportation: " + Transportation + "\n";
            output += "Duration: " + Duration + "\n";
            output += "Season: " + Season + "\n";
            output += "Accomodation: " + Accommodation + "\n";
            if (Hotel != null)
            {
                output += "Hotel: " + Hotel + "\n";
            }
            if (Similarity != 0)
            {
                output += "Similarity: " + Similarity + "\n";
            }

            return output;
        }

        public int CompareTo(object obj)
        {
            var other = obj as Case;
            if (other == null)
                return 0;
            if (Similarity == other.Similarity)
            {
                return 0;
            }
            return Similarity < other.Similarity ? 1 : -1;
        }
    }
}
